af_ktls.c

```diff
@@ -883,15 +885,17 @@ static int tls_sendmsg(struct socket *sock, struct msghdr *msg, size_t size)
 
 	tls_make_aad(tsk, 0, tsk->aad_send, size, tsk->iv_send);
 
-	ret = af_alg_make_sg(&tsk->sgl_send, &msg->msg_iter, size);
-	if (ret < 0)
-		goto send_end;
-
-	sg_unmark_end(&tsk->sgaad_send[1]);
-	sg_chain(tsk->sgaad_send, 2, tsk->sgl_send.sg);
-
-	sg_unmark_end(tsk->sgl_send.sg + tsk->sgl_send.npages - 1);
-	sg_chain(tsk->sgl_send.sg, tsk->sgl_send.npages + 1, tsk->sgtag_send);
+    while (iov_iter_count(&msg->msg_iter)) {
+        size_t seglen = iov_iter_count(&msg->msg_iter);
+        int len = af_alg_make_sg(&tsk->sgl_send[cnt], &msg->msg_iter, seglen);
+        if (len < 0)
+            goto send_end;
+        ret += len;
+        if (cnt)
+            af_alg_link_sg(&tsk->sgl_send[cnt-1], &tsk->sgl_send[cnt]);
+        iov_iter_advance(&msg->msg_iter, len);
+        cnt++;
+    }
```

Comment:
I think it would worth it to use ```kernel_sendpage()``` even for ```sendmsg(2)```/tls_sendmsg() to save copies (not only for tls_sendpage()). Something similar will be needed for that approach for sure (iterating over iovec), but we will allocate needed pages each time tls_sendmsg() will be called - that's why I suggested to wrap it to a single function to easily reuse it and not depend on ```tsk```.